torch/_dynamo/backends/distributed.py

```diff
@@ -322,7 +329,18 @@ def run_node(self, n: Node) -> Any:
             # Finally, we have to produce inputs for use compiling the next submodule,
             # and these need to be FakeTensors, so we execute the module under fake_mode
             with self.fake_mode:
-                return curr_submod(*new_args, **kwargs)
+                # TODO - make sure the first outputs of inductor match with user visible outputs
```

Comment:
@bdhirsh is this correct ?

Reply:
What I need is a mapping from dynamo output index -> aot autograd output index..

Reply:
Does aot autograd output here == the outputs of the forward FX graph that we send to inductor?

The calling convention is kind of complicated :/ ideally we would factor it into a shared util that we can re-use here, but it should look something like:

(*mutated_inputs_that_go_outside_the_graph, *fw_outs, *fw_intermediate_bases, *saved_tensors, *saved_symints)

Two more complications are:

(1) subclasses make this worse, because dynamo sees subclass outputs, but the aot graph sees plain tensor outputs. So if you have subclasses that desugar into multiple tensors, the calling convention gets more complicated (you need to __tensor_unflatten__ the "buckets" of inner tensors back into each subclass output). Code for that here:  https://github.com/pytorch/pytorch/blob/main/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py#L781

(2) @angelayi is working on adding effect tokens to functionalization, to properly functionalize torchbind object op calls. That will be another calling convention change (the effect tokens are extra tensor inputs/outputs in the graph that dynamo does not see.

Reply:
tbh - given all of the above, and given that ddp_optimizer is a "temporary" solution, I wonder if it would be safer to ban some of the above cases in ddp_optimizer rather than try to support all possible calling convention changes that AOTAutograd makes (and risk them diverging later).

Examples of things we could ban:

(1) subclasses (? probably ok, unclear if someone will come to us asking for ddp + subclasses to work before compiled ddp is more hardened)

(2) intermediates bases (pretty rare)

(3) input mutations (also... unclear? input mutations that we keep outside the graph are ideally relatively rare)